# Trabalho 1

## Setando o ambiente

Olá pessoal, para esse trabalho usaremos a base de dados da fapesp, para isso vocês precisam ter tudo configurado ai. Com o docker instalado no pc, rodem:

> Se vocês ja tiverem o docker configurado com a base de dados da fapesp, podem pular essa parte

```bash
docker compose up -d
```

O comando anterior so cria o database, O script de carga espera um banco chamado fapcov2103 então precisamos criar ele

```bash
docker compose exec -e PGPASSWORD=mysecretpassword db \
    psql -U myuser -d mydatabase -c "CREATE DATABASE fapcov2103;"
```

depois rode o script de carga:

```bash
docker compose exec -e PGPASSWORD=mysecretpassword db \
    psql -U myuser -d fapcov2103 -f /data/FapCov2103/@Scripts/load_covid.sql
```

se der erro boa sorte :D

In [5]:
# Codigo para validar se deu certo (So pra rodar, precisa entender nao)

import pandas.io.sql as pdsql
import pandas as pd
from sqlalchemy import create_engine, text

engine = create_engine('postgresql://myuser:mysecretpassword@localhost/fapcov2103')
conn = engine.connect()

query = text("""
    SELECT 'Pacientes' AS tabela, Count(*) FROM d2.pacientes
    UNION ALL
    SELECT 'ExamLabs', Count(*) FROM d2.examlabs
    UNION ALL
    SELECT 'Desfechos', Count(*) FROM d2.desfechos;
""")
pdsql.read_sql(query, conn)


,tabela,count
0,Pacientes,14673
1,ExamLabs,2952999
2,Desfechos,89937


## Entendendo o dataset

O esquema `D2` une os dados dos hospitais que **têm desfecho** registrado. Na base oficial completa da disciplina isso seria `HSL` + `HCSP`, mas no recorte de dados carregado neste repositório só o `HSL` tem arquivo de `Desfechos` (`HC` não tem) — então aqui, na prática, `D2` = `HSL`. Dá pra confirmar isso rodando `SELECT DISTINCT De_Hospital FROM d2.pacientes`.

Esse schema tem 3 tabelas. Toda a tarefa gira em torno de como elas se conectam.

---

### Pacientes

| ID_PACIENTE                      | IC_SEXO | AA_NASCIMENTO | CD_PAIS | CD_UF | CD_MUNICIPIO | CD_CEPREDUZIDO |
|-----------------------------------|---------|---------------|---------|-------|--------------|----------------|
| CFF55DA6AFD2DBF06851795AB62AD6CD | M       | 1970          | BR      | SP    | SAO PAULO    | CCCC           |

* `ID_PACIENTE` é um hash que representa o id os demais campos da pra entender kkk
* `CD_CEPREDUZIDO` e `CD_MUNICIPIO` vem como MMMM/CCCC quando são desconhecidos

---

### Desfechos

Um paciente pode ter vários atendimentos, um por linha

| ID_PACIENTE | ID_ATENDIMENTO | DT_ATENDIMENTO | DE_TIPO_ATENDIMENTO | DE_CLINICA      | DT_DESFECHO | DE_DESFECHO                 |
|-------------|-----------------|-----------------|----------------------|-----------------|-------------|------------------------------|
| 11B85F4A... | 713BB4E2...     | 26/02/2020      | Pronto Atendimento   | Clínica Médica  | 26/02/2020  | Desistência do atendimento  |

* Cada atendimento é uma passagem pelo hospital, pode ser consulta, internação, etc. Identificado pelo `ID_ATENDIMENTO`, chave dessa tabela é `(ID_PACIENTE, ID_ATENDIMENTO)`

---

### ExamLabs

* Nela um exame pode virar varias linhas

| ID_PACIENTE | ID_ATENDIMENTO | DT_COLETA  | DE_ORIGEM   | DE_EXAME               | DE_ANALITO              | DE_RESULTADO | CD_UNIDADE |
|-------------|-----------------|------------|-------------|--------------------------|---------------------------|--------------|------------|
| 9BB15EA1... | E7E88B60...     | 22/01/2021 | Hemodiálise | 17 Hidroxipregnenolona | 17-Hidroxi Pregnenolona | 23           | ng/dL      |

Um exame (`DE_EXAME`, ex: "Hemograma completo") pode medir vários analitos diferentes (`DE_ANALITO`, ex: hemoglobina, hematócrito, leucócitos,...). Isso significa que:

* Um único exame pedido gera várias linhas nessa tabela (uma por analito medido)
* `DE_RESULTADO` é texto livre: às vezes é um número ("23"), às vezes é uma categoria ("Negativa"), às vezes vem com espaços/pontuação estranha. **Isso é exatamente o problema que a tarefa quer resolver**
* A chave completa de um exame realizado é `(ID_PACIENTE, ID_ATENDIMENTO, DT_COLETA, DE_EXAME)` e mesmo assim, pode haver mais de um exame igual no mesmo dia, o que gera duplicatas dentro dessa própria chave

---

### Como isso se conecta com a tarefa?

Precisaremos criar uma tabela, essencialmente, uma linha por exame de covid/corona coletada enriquecida com:

* Os dados fixos do paciente (join com `Pacientes` por `ID_PACIENTE`)
* O desfecho mais recente do paciente (join com `DESFECHOS`, escolhendo o atendimento mais recente),
* E a classificação P/N/Brancho calculada a partir do texto sujo de `DE_RESULTADO`

---

# Entendendo o problema

**Objetivo:** criar e alimentar uma nova tabela na base `FapCov2103`, no esquema `D2`.

## Chave

A tabela deve ser identificada por `(Id_Paciente, Id_Atendimento, Dt_Coleta)`.

## Atributos dependentes

- todos os atributos do paciente;
- a data do atendimento;
- a data do desfecho;
- o valor do desfecho (`De_Desfecho`);
- o exame (`DE_Exame`);
- o resultado do exame (`DE_Resultado`);
- uma coluna calculada `Classe` (tipo `CHAR`), a partir do resultado do exame:
  - `'P'` se o resultado for positivo;
  - `'N'` se o resultado for negativo;
  - `' '` (branco) se não for possível identificar.

## Seleção das tuplas

Devem entrar na tabela as linhas de exames de **'Covid'** ou **'Corona'**, mas só de pacientes que **tenham algum desfecho registrado**.

## Limpeza de dados

O enunciado avisa que preparar os dados pode exigir descartar parte deles, e impõe um critério mensurável pra `Classe`:

> pelo menos **95%** das tuplas devem ter valor `'N'` ou `'P'` (o resto pode ficar em branco).

É esse número que vamos validar ao longo do notebook, depois de tratar as diversas representações do resultado do exame.

---

## Parte 1 - Explorar DE_Exame e DE_Resultado

In [ ]:
#vamos precisar criar um CASE da coluna Classe, mas antes precisamos ver quais valores reais que existem

"""
Apresentando De_exame e contagem 
da tabela de examlabs 
onde o DE_Exame é covid ou corona 
agrupado por DE_Exame (para gerar o Count) 
Ordenado
"""

query = text("""
    SELECT DE_Exame, Count(*) AS qtd
    FROM d2.examlabs
    WHERE DE_Exame ~* 'covid|corona'
    GROUP BY DE_Exame
    ORDER BY qtd DESC;
""")

pdsql.read_sql(query, conn)


,de_exame,qtd
0,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",16206
1,COVID-19-Sorologia IgM e IgG por quimiluminesc...,11196
2,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,10614
3,"COVID-19-Teste Rápido (IgM e IgG), soro",515
4,"Sorologia - Coronavírus, IgG",304
5,"Sorologia - Coronavírus, IgA",242
6,"COVID-19, anticorpos IGA e IGG, soro",30


In [ ]:
# Fazendo o mesmo para o DE_Resultado
pd.set_option('display.max_rows', 200) # 200 linhas para a exibição de um DataFrame do Pandas antes que ele seja truncado.


"""
Igual o DE_Exame
"""


query = text("""
    SELECT DE_Resultado, Count(*) AS qtd
    FROM d2.examlabs
    WHERE DE_Exame ~* 'covid|corona'
    GROUP BY DE_Resultado
    ORDER BY qtd DESC;
""")

pdsql.read_sql(query, conn)


,de_resultado,qtd
0,DETECTADO,12144
1,DETECTADO (POSITIVO),3767
2,swab de nasofaringe,3765
3,raspado de material do trato respiratório,2907
4,NÃO REAGENTE,2286
...,...,...
830,"88,4",1
831,"88,5",1
832,"8,86",1
833,"89,4",1


### O que os dados mostram

* DE_Exame ficou limpo, com 7 tipos de exame, todos de Covid, logo o filtro está correto e sem falso positivo

* DE_Resultado TEM 835 valores distintos, isso pois cada DE_Exame mede varios DE_Analito diferentes, e nem todo analito é o "resultado"

In [10]:

# A ideia é descobrir para cada tipo de exame Covid/Corona, quais analitos ele mede. 
# Dentro de cada DE_Exame, quantas linhas cada DE_Analito gera e quais são esses analitos

#Sem essa query os 835 valores de DE_Resultado pareceriam aleatórios. Com ela da pra ver o padrão: Um mesmo exame gerando várias linhas por natureza diferente
query = text("""
    SELECT DE_Exame, DE_Analito, Count(*) AS qtd
    FROM d2.examlabs
    WHERE DE_Exame ~* 'covid|corona'
    GROUP BY DE_Exame, DE_Analito
    ORDER BY DE_Exame, qtd DESC;
""")
pdsql.read_sql(query, conn)


,de_exame,de_analito,qtd
0,"COVID-19, anticorpos IGA e IGG, soro","Covid 19, Anticorpos IgA",5
1,"COVID-19, anticorpos IGA e IGG, soro","Covid 19, Anticorpos IgA e IgG, Interpretação",5
2,"COVID-19, anticorpos IGA e IGG, soro","Covid 19, Anticorpos IgA, Interpretação",5
3,"COVID-19, anticorpos IGA e IGG, soro","Covid 19, Anticorpos IgG",5
4,"COVID-19, anticorpos IGA e IGG, soro","Covid 19, Anticorpos IgG, Interpretação",5
5,"COVID-19, anticorpos IGA e IGG, soro","Covid 19, observação",5
6,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",Coronavírus (2019-nCoV),8207
7,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",Material (2019-nCoV),7995
8,"COVID-19-PCR para SARS-COV-2, Vários Materiais...","Covid 19, Observação",4
9,COVID-19-Sorologia IgM e IgG por quimiluminesc...,"Covid 19, Anticorpos IgG, Quimioluminescência",1546


## PARTE 2 - Construindo a regra de classificação (CLASSE)

* Vamos criar uma classe para validar os tipos de de DE_Resultados como Positivo ou Negativo

* Como ja vimos "NÃO REAGENTE" contém a palavra "REAGENTE", a ordem dos WHEN importa, precisamos chegar os padrões negativos primeiro, se não "NÃO REAGENTE cairia em positivo por engano"



In [ ]:
# Regex usados explicados:

# n[ãa]o\s*(detectad|reagente) -> pega NÃO detectado/detecta reagente -> 'N'
# negativ pega Negativo/Negativa/NEGATIVO -> 'N'
# detectad | positiv | reagente -> so chega aqui se nao caiu no negativo antes de -> 'P'


#Idea é dar uma CLASSE para cada DE_Resultado como positivo/negativo ou se nao é possivel analisar
#Para isso podemos usar a estrutura CASE WHEN THEN.

#Sem o regex NAO REAGENTE cairia tanto no positivo quanto no negativo por exemplo

query = text("""
    SELECT
        CASE
            WHEN DE_Resultado ~* 'n[ãa]o\\s*(detectad|reagente)' OR DE_Resultado ~* 'negativ' THEN 'N'
            WHEN DE_Resultado ~* 'detectad|positiv|reagente' THEN 'P'
            ELSE ' '
        END AS classe,
        Count(*) AS qtd
    FROM d2.examlabs
    WHERE DE_Exame ~* 'covid|corona'
    GROUP BY classe
    ORDER BY qtd DESC;
""")
df = pdsql.read_sql(query, conn)
df['pct'] = 100 * df['qtd'] / df['qtd'].sum()
df


,classe,qtd,pct
0,P,18956,48.472141
1,,13344,34.121768
2,N,6807,17.406091


Boa parte do branco é estrutural e não erro da lógica do regex, são linhas de `Material/Observações` (tipos de amostra) e índices númericos de anticorpo que nunca serão P/N, porque não são o resultado, são metadado do mesmo exame!

A ideia é entender para cada chave `(ID_Paciente, Id_Atendimento, Dt_Coleta)`, vamos escolher 1 linha só, priorizando a que já deu P/N sobre as que deram branco. É só depois desse descarte que a % de 95% deve ser medida (O enunciado pede que deve ser tratadas as diversasa representações do resultado do exame, de maneira que pelo menos 95% das tuplas tenham valor "N" ou "P"), assim a métrica que temos até aqui não é valida ainda pois foi calculada em cima desses metadados (material, observacao, indices numericos) ainda misturadas que nem deveriam virar linha na tabela final

## PARTE 3 - Escolher 1 linha por chave

A ferramente `DISTINC ON` do Postgres. Ele agrupa por umas colunas e mantém só a primeira linha de cada grupo, segundo a ordem que voce definir no `ORDER BY`.

A estratégia é para cada `(Id_Paciente, Id_Atendimento, Dt_Coleta)`ordenar as linhas candidatas colocando as que viraram `P` / `N` primeiro, e as em branco por último 

In [ ]:
"""
1. Subquery interna do FROM (sub):
- É exatamente a parte 2, pegando todas as linhsa cruas de exame de covid e calculando a casse de cada uma, sem agrupar nada ainda
2. DISTINCT ON (Id_Paciente, Id_Atendiment, Dt_Coleta)
- Agrupo as linhas por essa condição de 3 colunas, e me devolva só a 1 linha de cada grupo
- A linha escolhida é definida pelo ORDER BY
3. ORDER BY:
- as 3 primeiras colunas tem que bater com as do DISTINCT ON (se nao da erro)
- Cria um "0" para as linhas que ja sao P ou N e "1" para as linhas em branco. Como o order by é crescente vai fazer começar em 0. 
- Se dentro do mesmo grupo tiver por exemplo, duas linhas ambas com P (paciente fez 2 exames de covid diferentes no mesmo dia e ambos positivos) o Classe depois faz o desempate onde decide: 'N' < 'P' alfabeticamente, logo da prioridade ao N e P e brancos são descartados
"""

query = text("""
    SELECT DISTINCT ON (Id_Paciente, Id_Atendimento, Dt_Coleta)
        Id_Paciente, Id_Atendimento, Dt_Coleta, DE_Exame, DE_Resultado, Classe
    FROM (
        SELECT *,
            CASE
                WHEN DE_Resultado ~* 'n[ãa]o\\s*(detectad|reagente)' OR DE_Resultado ~* 'negativ' THEN 'N'
                WHEN DE_Resultado ~* 'detectad|positiv|reagente' THEN 'P'
                ELSE ' '
            END AS Classe
        FROM d2.examlabs
        WHERE DE_Exame ~* 'covid|corona'
    ) sub
    ORDER BY Id_Paciente, Id_Atendimento, Dt_Coleta,
             CASE WHEN Classe IN ('P','N') THEN 0 ELSE 1 END,  -- P/N antes de branco
             Classe;                                            -- desempate estável
""")
exames_dedup = pdsql.read_sql(query, conn)
exames_dedup


,id_paciente,id_atendimento,dt_coleta,de_exame,de_resultado,classe
0,000150DB429AFF026ECD130B0A076CEF,644998DCF29A2F3EF3AA0EC3D7157666,2020-12-03,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
1,00017961865C4F766FDBB3CD8FE0BFB0,B57529234B6132E278264F273ADB99DC,2020-08-25,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
2,0003B368F65004E14332CD44BEE6E600,1E774FD413636840E7CAAE47817F8120,2020-12-28,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
3,000F0BC139D2846DB86AA32B8F05B215,9D6B251913C9B614DCC7B6518F297C75,2020-08-28,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
4,00177290279939FB33386B29198C450E,94B6DCEC9250DDA79934306B7C214F81,2020-06-11,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
...,...,...,...,...,...,...
20079,FFEC3898BAA04751EB00C108270B8F7E,B3EC126090E23983F3552ABB3BAD42D2,2020-06-23,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
20080,FFEE2A236442ECF6D9042F39A8EA73FD,10F46E547DC0EA8A39F887FDF25A9228,2020-12-06,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
20081,FFEE2A236442ECF6D9042F39A8EA73FD,26E534111888FACABE6B25B96E4F75A6,2021-01-23,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
20082,FFF5753408C98D5E0218931420B6AF85,68ADE28F0236E35E29CABC1F12622CBE,2020-09-27,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P


Reduziu de 39 mil linhas para 20 mil linhas. Confirmando que foi removido muito metadado duplicado

Faltando apenas medir a metrica real

In [14]:
#Metrica do enunciado, feito em pandas ao invés de SQL
#é pego a coluna classe, transforma em True se for P ou N e como True conta como 1, tirar a média desses valores é exatamente
#verificar a proporção de linhas que são P|N. Ai múltiplicamos por 100 para ter a porcentagem

pct = 100 * exames_dedup['classe'].isin(['P', 'N']).mean() 
pct


np.float64(99.21330412268472)

99,21% passaram, o enunciado exige 95% então estamos no caminho certo!

## PARTE 4 - Trazer os dados de Pacientes e Desfechos

Até aqui só trabalhamos em cima de ExamLabs: filtramos covid e reduzimos para 1 linha por chave `(Id_Paciente, Id_Atendimento, Dt_Coleta)`, com a `Classe`calculada. Isso resolve a parte de seleção das tuplas e da coluna calculada

Mas a tabela final também precisa de atributos dependentes que ainda não trouxemos (olhando o enunciado):
* Todos os atributos do paciente -> Vem de paciente
* A data do atendimento, a data do desfecho, e o valor do desfecho -> Vem de Desfechos

Precisamos portanto enriquecer cada linha de exame que já temos com esses dois pedaços de informaçãom, via JOIN

Por paciente é trivial o Join, porém por Desfecho temos alguns pontos a se atentar

O desfecho não tem 1 linha por paciente, tem N linhas por paciente, uma por cada atendimento que ele teve (pronto-socorro, internação, etc), cada um com seu próprio De_Desfecho (alta, óbito, transferencia) e sua própria data.

Só que o enunciado pede pra tabela final ter, como atributo dependente, o "o desfecho do paciente" e ai precisamos escolher qual dos desfechos

são duas opções:

1. Casar pelo atendimento exato do exame: usando o desfecho registrado no mesmo Id_Atendimento em que o exame de Covid foi coletado
2. Pegar o desfecho mais recente do paciente: Olhando todos os atendimentos daquele paciente e usar o de data mais recente, independente de ter sido o mesmo atendimento do exame Covid.

Validando:

In [16]:
# lista os pareces (Id_paciente, Id_Atendimento) onde foi realizado exame de covid
# e conta quantos pares existem no total na primeira coluna
# ja na segunda coluna, quanto desses pares tem uma linha de Desfechos com o mesmo Id_paciente e Id_Atendimento (opcao 1)
# A terceiro coluna conta quantos tem algum Id_Paciente igual em Desfechos, não importando o atendimento (Opção 2: Join por paciente)

query = text("""
    WITH covid AS (
        SELECT DISTINCT Id_Paciente, Id_Atendimento
        FROM d2.examlabs
        WHERE DE_Exame ~* 'covid|corona'
    )
    SELECT
        Count(*) AS total_pares_paciente_atendimento,
        Count(*) FILTER (WHERE EXISTS (
            SELECT 1 FROM d2.desfechos D
            WHERE D.Id_Paciente = covid.Id_Paciente AND D.Id_Atendimento = covid.Id_Atendimento
        )) AS com_desfecho_mesmo_atendimento,
        Count(*) FILTER (WHERE EXISTS (
            SELECT 1 FROM d2.desfechos D WHERE D.Id_Paciente = covid.Id_Paciente
        )) AS com_desfecho_paciente_qualquer
    FROM covid;
""")
pdsql.read_sql(query, conn)


,total_pares_paciente_atendimento,com_desfecho_mesmo_atendimento,com_desfecho_paciente_qualquer
0,19370,19253,19369


Daqui é possivel ver uma diferença real (117 exames) entre "Casar pelo atendimento exato" e "casar por paciente", pegando qualquer desfecho dele. Logo a escolha do Join importa (tem 117 exames de pacientes reais no jogo)

Daqui concluimos que compensa juntar Desfechos por Id_Paciente (nao por Id_Atendimento exato), pq cobre praticamente todos os casos. E no enunciado tbm fala isso ("Pacientes que tenham algum desfecho")

A ideia é aproveitar a mesma lógica que fizemos no Passo 3 para agrupar

In [18]:
# Agruparemos por Id_Paciente 
# Queremos manter, dentro de cada paciente a linha do atendimento mas recente, então o critério de ordenação
# decide. Caso 2 pacientes tenham a msm data a escolha vira arbitrária ai acrescenta o Id_Atendimento como critério extra

# assim, trazemos depois Data do atendimento, data do desfecho, valor do desfecho (se vier como DDMMAA converte para NULL)

query = text("""
    SELECT DISTINCT ON (Id_Paciente)
        Id_Paciente,
        Dt_Atendimento AS Dt_FAtendimento,
        CASE WHEN Dt_Desfecho ~ 'DDMMAA' THEN NULL
             ELSE TO_DATE(Dt_Desfecho, 'DD/MM/YYYY') END AS Dt_FDesfecho,
        De_Desfecho AS De_FDesfecho
    FROM d2.desfechos
    ORDER BY Id_Paciente, Dt_Atendimento DESC, Id_Atendimento;
""")
desfecho_final = pdsql.read_sql(query, conn)
desfecho_final


,id_paciente,dt_fatendimento,dt_fdesfecho,de_fdesfecho
0,000150DB429AFF026ECD130B0A076CEF,2020-12-03,2020-12-03,Alta Administrativa
1,00017961865C4F766FDBB3CD8FE0BFB0,2020-08-25,2020-08-25,Alta médica melhorado
2,0003B368F65004E14332CD44BEE6E600,2020-12-28,2020-12-28,Alta médica melhorado
3,000F0BC139D2846DB86AA32B8F05B215,2021-06-09,2021-06-09,Alta Administrativa
4,00177290279939FB33386B29198C450E,2020-06-13,2020-06-13,Alta médica melhorado
...,...,...,...,...
14667,FFE724271952607EE4DA68C847297386,2021-05-13,2021-05-13,Alta Administrativa
14668,FFEC3898BAA04751EB00C108270B8F7E,2021-04-28,2021-04-29,Alta médica melhorado
14669,FFEE2A236442ECF6D9042F39A8EA73FD,2021-02-03,2021-02-03,Alta médica melhorado
14670,FFF5753408C98D5E0218931420B6AF85,2020-11-17,2020-11-17,Alta médica melhorado


In [ ]:
#pega desfecho final , se for nulo vira true e depois soma!
desfecho_final['dt_fdesfecho'].isna().sum() 

np.int64(164)

São  164 caso como `dt_fdesfecho` ainda como NULL. Provavelmente porque parte desses casos não terminaram ainda na hora de definir o dataset

----

Certo! com isso definido podemos juntar o que vimos

In [ ]:
"""
    ---
    WHITH exame covid AS (...)  - PARTE 3 (p1)
        desfecho final AS (...)  - DEFINIDO NO COMEÇO DA PARTE 4 (p2)
    SELECT ...                   - JUNTA OS DOIS + PACIENTE (p3)
    FROM exames_covid E
    JOIN d2 paciente P ON ...
    JOIN desfecho_final F ON ...
    ----

    explicando p3:
    A lógica aqui é E, P e DF são aliases das 3 fontes. Logo cada coluna do select é escolhida de onde ela realmente pertence: ex - Atributos do paciente vem de P (d2.pacientes)


    Logo o que essa query faz é:
    1. Pega a lista de exames covid/corona ja reduzidas a 1 por chave (paciente, atendimento, data), com a Classe calculada
    2. Pega, separadamente o desfecho mais recente de cada paciente
    3. Junta essasa duas listas com a tabela pacientes, casando tudo pelo Id_Paciente
    4. Como os dois JOIN são "estritos" (nao LEFT JOIN), qualquer exame de um paceiente que não tenha desfecho registrado é descartado automaticamente, é assim que a regrea "só pacientes que tem desfecho" é aplicada
    5. O resultado final é uma linha por exame de covid coletado, já com todos os dados do paciente e do desfecho colados do lado, pronta para virar a tabela definitiva
"""





query = text("""
    WITH exames_covid AS (
        SELECT DISTINCT ON (Id_Paciente, Id_Atendimento, Dt_Coleta)
            Id_Paciente, Id_Atendimento, Dt_Coleta, DE_Exame, DE_Resultado, Classe
        FROM (
            SELECT *,
                CASE
                    WHEN DE_Resultado ~* 'n[ãa]o\\s*(detectad|reagente)' OR DE_Resultado ~* 'negativ' THEN 'N'
                    WHEN DE_Resultado ~* 'detectad|positiv|reagente' THEN 'P'
                    ELSE ' '
                END AS Classe
            FROM d2.examlabs
            WHERE DE_Exame ~* 'covid|corona'
        ) sub
        ORDER BY Id_Paciente, Id_Atendimento, Dt_Coleta,
                 CASE WHEN Classe IN ('P','N') THEN 0 ELSE 1 END,
                 Classe
    ),
    desfecho_final AS (
        SELECT DISTINCT ON (Id_Paciente)
            Id_Paciente,
            Dt_Atendimento AS Dt_FAtendimento,
            CASE WHEN Dt_Desfecho ~ 'DDMMAA' THEN NULL
                 ELSE TO_DATE(Dt_Desfecho, 'DD/MM/YYYY') END AS Dt_FDesfecho,
            De_Desfecho AS De_FDesfecho
        FROM d2.desfechos
        ORDER BY Id_Paciente, Dt_Atendimento DESC, Id_Atendimento
    )
    SELECT
        E.Id_Paciente, E.Id_Atendimento, E.Dt_Coleta,
        P.Ic_Sexo, P.Aa_Nascimento, P.Cd_Pais, P.Cd_Uf, P.Cd_Municipio, P.Cd_Cepreduzido,
        DF.Dt_FAtendimento AS Dt_Atendimento,
        DF.Dt_FDesfecho    AS Dt_Desfecho,
        DF.De_FDesfecho    AS De_Desfecho,
        E.DE_Exame,
        E.DE_Resultado,
        E.Classe
    FROM exames_covid E
    JOIN d2.pacientes P     ON P.Id_Paciente = E.Id_Paciente
    JOIN desfecho_final DF ON DF.Id_Paciente = E.Id_Paciente;
""")
tabela_final = pdsql.read_sql(query, conn)
tabela_final


,id_paciente,id_atendimento,dt_coleta,ic_sexo,aa_nascimento,cd_pais,cd_uf,cd_municipio,cd_cepreduzido,dt_atendimento,dt_desfecho,de_desfecho,de_exame,de_resultado,classe
0,000150DB429AFF026ECD130B0A076CEF,644998DCF29A2F3EF3AA0EC3D7157666,2020-12-03,F,1993.0,BR,SP,SAO PAULO,CCCC,2020-12-03,2020-12-03,Alta Administrativa,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
1,00017961865C4F766FDBB3CD8FE0BFB0,B57529234B6132E278264F273ADB99DC,2020-08-25,M,1967.0,BR,SP,NaN,CCCC,2020-08-25,2020-08-25,Alta médica melhorado,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
2,0003B368F65004E14332CD44BEE6E600,1E774FD413636840E7CAAE47817F8120,2020-12-28,F,1968.0,BR,SP,SAO PAULO,CCCC,2020-12-28,2020-12-28,Alta médica melhorado,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
3,000F0BC139D2846DB86AA32B8F05B215,9D6B251913C9B614DCC7B6518F297C75,2020-08-28,M,1980.0,BR,SP,SAO PAULO,CCCC,2021-06-09,2021-06-09,Alta Administrativa,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
4,00177290279939FB33386B29198C450E,94B6DCEC9250DDA79934306B7C214F81,2020-06-11,F,1982.0,BR,SP,SAO PAULO,CCCC,2020-06-13,2020-06-13,Alta médica melhorado,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20078,FFEC3898BAA04751EB00C108270B8F7E,B3EC126090E23983F3552ABB3BAD42D2,2020-06-23,M,1982.0,BR,DF,BRASILIA,CCCC,2021-04-28,2021-04-29,Alta médica melhorado,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
20079,FFEE2A236442ECF6D9042F39A8EA73FD,10F46E547DC0EA8A39F887FDF25A9228,2020-12-06,F,1986.0,BR,SP,SAO PAULO,CCCC,2021-02-03,2021-02-03,Alta médica melhorado,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
20080,FFEE2A236442ECF6D9042F39A8EA73FD,26E534111888FACABE6B25B96E4F75A6,2021-01-23,F,1986.0,BR,SP,SAO PAULO,CCCC,2021-02-03,2021-02-03,Alta médica melhorado,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P
20081,FFF5753408C98D5E0218931420B6AF85,68ADE28F0236E35E29CABC1F12622CBE,2020-09-27,F,2004.0,BR,SP,SAO PAULO,CCCC,2020-11-17,2020-11-17,Alta médica melhorado,Detecção de Coronavírus (NCoV-2019) POR PCR (A...,DETECTADO,P


In [24]:
# Validando se a chave continua unica
tabela_final.shape[0] == tabela_final[['id_paciente', 'id_atendimento', 'dt_coleta']].drop_duplicates().shape[0]


True

In [25]:
# Validando se o % de P/N se mantem  >= 95% (o join nao deveria mexer na classe)
100 * tabela_final['classe'].isin(['P', 'N']).mean()


np.float64(99.21326495045561)

In [27]:
# Quantas linhas tem dt_desfecho:
tabela_final['dt_desfecho'].isna().sum()


np.int64(320)

## PARTE 5 - Materializar a tabela em D2

Até aqui `tabela_final` só existe em memória (DataFrame do pandas). Precisamos de uma tabela real no banco, no schema `D2` — por isso trocamos `SELECT` por `CREATE TABLE ... AS SELECT`, que já cria a tabela populada.

Três detalhes importantes no código abaixo:

1. **`::CHAR(1)` na `Classe`** — o enunciado exige explicitamente que essa coluna seja do tipo `CHAR`. Até agora ela saía como texto genérico; aqui forçamos o tipo certo.
2. **`ADD PRIMARY KEY` depois de criar a tabela** — é uma prova mais forte de que a chave é única do que a checagem que fizemos em pandas (`drop_duplicates`): se existisse qualquer duplicata, o próprio Postgres **recusaria** criar essa chave e o código pararia com erro.
3. **`conn.commit()` no final** — sem isso, `CREATE TABLE` e `ALTER TABLE` ficam pendurados numa transação nunca confirmada e não persistem no banco (mesmo problema de transação que vimos antes, quando um erro deixou a conexão travada).

In [ ]:
table_name = "exames_covid_desfecho" 

# Remove a tabela se já existir, pra poder rodar essa célula de novo sem erro
conn.execute(text(f"DROP TABLE IF EXISTS d2.{table_name};"))

# Materializa a query da Parte 4 como tabela real (CREATE TABLE ... AS)
conn.execute(text(f"""
    CREATE TABLE d2.{table_name} AS
    WITH exames_covid AS (
        SELECT DISTINCT ON (Id_Paciente, Id_Atendimento, Dt_Coleta)
            Id_Paciente, Id_Atendimento, Dt_Coleta, DE_Exame, DE_Resultado, Classe
        FROM (
            SELECT *,
                CASE
                    WHEN DE_Resultado ~* 'n[ãa]o\\s*(detectad|reagente)' OR DE_Resultado ~* 'negativ' THEN 'N'
                    WHEN DE_Resultado ~* 'detectad|positiv|reagente' THEN 'P'
                    ELSE ' '
                END::CHAR(1) AS Classe
            FROM d2.examlabs
            WHERE DE_Exame ~* 'covid|corona'
        ) sub
        ORDER BY Id_Paciente, Id_Atendimento, Dt_Coleta,
                 CASE WHEN Classe IN ('P','N') THEN 0 ELSE 1 END,
                 Classe
    ),
    desfecho_final AS (
        SELECT DISTINCT ON (Id_Paciente)
            Id_Paciente,
            Dt_Atendimento AS Dt_FAtendimento,
            CASE WHEN Dt_Desfecho ~ 'DDMMAA' THEN NULL
                 ELSE TO_DATE(Dt_Desfecho, 'DD/MM/YYYY') END AS Dt_FDesfecho,
            De_Desfecho AS De_FDesfecho
        FROM d2.desfechos
        ORDER BY Id_Paciente, Dt_Atendimento DESC, Id_Atendimento
    )
    SELECT
        E.Id_Paciente, E.Id_Atendimento, E.Dt_Coleta,
        P.Ic_Sexo, P.Aa_Nascimento, P.Cd_Pais, P.Cd_Uf, P.Cd_Municipio, P.Cd_Cepreduzido,
        DF.Dt_FAtendimento AS Dt_Atendimento,
        DF.Dt_FDesfecho    AS Dt_Desfecho,
        DF.De_FDesfecho    AS De_Desfecho,
        E.DE_Exame,
        E.DE_Resultado,
        E.Classe
    FROM exames_covid E
    JOIN d2.pacientes P     ON P.Id_Paciente = E.Id_Paciente
    JOIN desfecho_final DF ON DF.Id_Paciente = E.Id_Paciente;
"""))

# Se sobrar qualquer duplicata na chave, essa linha falha -- é o teste definitivo de unicidade
conn.execute(text(f"""
    ALTER TABLE d2.{table_name}
        ADD PRIMARY KEY (Id_Paciente, Id_Atendimento, Dt_Coleta);
"""))

# Sem commit, nada disso persiste no banco
conn.commit()
print("Tabela criada e chave primária validada com sucesso!")


TUDO CERTO!

In [29]:
query = text("""
    SELECT
        Count(*) AS total_linhas,
        Count(*) FILTER (WHERE Classe IN ('P','N')) AS pn,
        Round(100.0 * Count(*) FILTER (WHERE Classe IN ('P','N')) / Count(*), 2) AS pct_pn
    FROM d2.exames_covid_desfecho;
""")
pdsql.read_sql(query, conn)


,total_linhas,pn,pct_pn
0,20083,19925,99.21


In [ ]:
query = text("SELECT * FROM d2.exames_covid_desfecho LIMIT 1;") #retornando uma query de exemplo
pdsql.read_sql(query, conn)


,id_paciente,id_atendimento,dt_coleta,ic_sexo,aa_nascimento,cd_pais,cd_uf,cd_municipio,cd_cepreduzido,dt_atendimento,dt_desfecho,de_desfecho,de_exame,de_resultado,classe
0,000150DB429AFF026ECD130B0A076CEF,644998DCF29A2F3EF3AA0EC3D7157666,2020-12-03,F,1993,BR,SP,SAO PAULO,CCCC,2020-12-03,2020-12-03,Alta Administrativa,"COVID-19-PCR para SARS-COV-2, Vários Materiais...",DETECTADO (POSITIVO),P
